Avoid heap walk and GC for MemberName fix-up on class redefinition

jdmpapin opened the pull request:
In class redefinition the affected `MemberName` objects were previously identified by iterating over all objects in the heap. The heap walk also necessitated a GC cycle to eliminate dead objects, which could otherwise cause the fix-up to crash.

This heap walk and GC cycle were causing a significant performance regression in redefinition for builds using OpenJDK `MethodHandle`s (i.e. Java 17+) vs. ones using J9 `MethodHandle`s. The regression was especially drastic in fast HCR mode, where no other heap walk is necessary and therefore the time taken for redefinition should be largely independent of the number of objects in the heap.

Now instead each `J9Class` has a list of JNI weak global references to all of the `MemberName` objects that are resolved (i.e. have `vmtarget` set) to a member of that class (as determined by the `clazz` field). Redefinition uses these lists to identify the affected `MemberName`s without inspecting any other objects, and in particular without walking the heap. Because the heap walk is no longer needed, neither is the GC cycle, and both are eliminated.

The use of weak references ensures that these lists don't prevent `MemberName` objects from being reclaimed, and in particular that they don't prevent class unloading.

When a `MemberName` is reclaimed, its corresponding weak reference is cleared, and both that reference and its linked list node become unnecessary. To avoid leaking this memory, such entries are removed just prior to the addition of a `MemberName`, but only if it can be expected that there are any. To this end, the VM now provides an implementation of a native method `MethodHandleNatives.markClassForMemberNamePruning()`, which is expected to (exist and) be called when a `MemberName` that is on a per-class list becomes unreachable (really, phantom reachable), such as from a cleaning action or finalizer.

Without this adjustment to the Java sources, the mechanism implemented in this commit will still work, but the `MemberName` list will leak memory if many `MemberName`s for the same class are created and then reclaimed.

Issue: #19152 
Issue: #19153

jdmpapin commented:
@tajila, could you please review or recommend a reviewer?

tajila commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -83,6 +83,79 @@ isPolymorphicMHMethod(J9JavaVM *vm, J9Class *declaringClass, J9UTF8 *methodName)
+		if (NULL != weakRef) {
+			vmFuncs->j9jni_deleteGlobalRef((JNIEnv*)currentThread, weakRef, JNI_TRUE);
+		}
+		vmFuncs->setCurrentExceptionUTF(currentThread, J9_EX_OOM_OS_HEAP | J9VMCONSTANTPOOL_JAVALANGOUTOFMEMORYERROR, NULL);
```
you can use `vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);`

tajila commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -3406,6 +3419,12 @@ typedef struct J9ArrayClass {
+	/* A linked list of weak global references to every resolved MemberName whose clazz is this class. */
+	J9MemberNameListNode *memberNames;
+	/* True if a MemberName object belonging to this class (so likely in memberNames) has been finalized. */
+	BOOLEAN needToPruneMemberNames;
```
instead of using a whole word, I think we still have some room in clazz->classFlags

babsingh commented:
Still taking a look at the changes. Also, adding @fengxue-IS to review the PR.

babsingh commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -1089,14 +1165,16 @@ Java_java_lang_invoke_MethodHandleNatives_resolve(
-				J9VMJAVALANGINVOKEMEMBERNAME_SET_CLAZZ(currentThread, membernameObject, new_clazz);
-				J9OBJECT_U64_STORE(currentThread, membernameObject, vm->vmindexOffset, (U_64)vmindex);
-				J9OBJECT_U64_STORE(currentThread, membernameObject, vm->vmtargetOffset, (U_64)target);
+				if (addMemberNameToClass(currentThread, membernameObject, new_clazz)) {
```
We are adding to the list from three different locations: init, resolve and expand. @fengxue-IS Can you confirm is the answer to the below questions is correct?
- Will we catch all MemberNames that are created from these three locations? **Yes**
- Can duplicate entries be added to the list? **No**

jdmpapin replied:
Maybe my description of "resolved" `MemberName`s was too loose. What we really need - and what this code adds to the list - are those `MemberName` instances that have `vmtarget`/`vmindex` set based on a particular member of some class. (It's `vmtarget` especially that's important, though `vmindex` also matters when using the extensions.) From grepping, these are the only two locations I found where these fields are stored to (except for `prepareToFixMemberNames()` and `fixMemberNames()`, which temporarily repurpose them, and then only in objects where they've already been set.) It's my impression that this corresponds more or less to what the Java source considers to be a resolved `MemberName`, but I haven't checked that it corresponds exactly

fengxue-IS replied:
There is an other place that sets the MemberName.clazz field in `Java_java_lang_invoke_MethodHandleNatives_expand`. Though that path doesn't really modify the dependency. For now I think its safe to assume all resolved MN should be caught by these two locations

fengxue-IS replied:
There is an other place that sets the MemberName.clazz field in Java_java_lang_invoke_MethodHandleNatives_expand. Though that path doesn't really modify the dependency. For now I think its safe to assume all resolved MN should be caught by these two locations

jdmpapin replied:
If `clazz` is set without `vmtarget`/`vmindex`, then no fixup is necessary

jdmpapin commented:
Rebased. This is just a clean rebase with no other changes. I'll update to address review comments next

fengxue-IS reviewed (changes requested):
Having a global `memberNameListsMutex` feels like a possible perf issue to me in this change.

few points to consider:
1. class specific lock would solve most of the collision issue.
2. rather than setting a `needToPruneMemberNames`, directly doing the pruning in finalization makes more sense.
3. can we add a `prev/next` hidden field to MemberName class and directly use those fields to create a double linked list? This avoid the need for JNIGlobalRef and  J9MemberNameListNode pool. It also allows more atomic add/remove which eliminate need for mutex.

jdmpapin commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -83,6 +83,79 @@ isPolymorphicMHMethod(J9JavaVM *vm, J9Class *declaringClass, J9UTF8 *methodName)
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9Class *j9clazz = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, clazzObject);
+
+	jobject weakRef = vmFuncs->j9jni_createGlobalRef((JNIEnv*)currentThread, memberNameObject, JNI_TRUE);
```
I'd prefer not to obfuscate the code to optimize for a highly unlikely error case

jdmpapin commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -83,6 +83,79 @@ isPolymorphicMHMethod(J9JavaVM *vm, J9Class *declaringClass, J9UTF8 *methodName)
+
+	jobject weakRef = vmFuncs->j9jni_createGlobalRef((JNIEnv*)currentThread, memberNameObject, JNI_TRUE);
+
+	omrthread_monitor_enter(vm->memberNameListsMutex);
```
I was asked to use a pool to allocate the list nodes. Since the pool needs a global mutex anyway, I used it to protect the lists as well. Outside of STW, the critical section will usually be constant-time, only taking longer if pruning is necessary

jdmpapin commented:
> rather than setting a `needToPruneMemberNames`, directly doing the pruning in finalization makes more sense.

The strategy using a flag can batch pruning passes. If some large number *n* of `MemberName` instances are reclaimed for a single class, then we could potentially get away with only one pruning pass over the list. But even if we don't, it seems likely that we'd be able to do much fewer than the *n* passes that would be necessary if we were to prune eagerly in the finalizer

> can we add a `prev/next` hidden field to MemberName class and directly use those fields to create a double linked list?

This requires more GC integration. A few of us have discussed designs like this, and for the time being we agreed to use these native lists of JNI weak refs for expediency so that we can address the problem in 0.44

jdmpapin commented:
Updated to call `vmFuncs->setNativeOutOfMemoryError` and to use a new flag in `classFlags` rather than adding an entire `BOOLEAN` to `J9Class`

I noticed that in `freeJavaVM()` I had accidentally put my cleanup within `#if defined(J9VM_OPT_CRIU_SUPPORT)`, so I've corrected that as well

tajila reviewed (approved):
LGTM. Just want to point out that this design is intended for 0.44 only. Due to the urgency we need to accept some sub-optimal behaviours for the time being.

babsingh commented:
jenkins test sanity,sanity.openjdk amac jdk11

babsingh commented:
jenkins test sanity,sanity.openjdk alinux jdk21

babsingh commented:
jenkins compile win jdk21

jdmpapin commented:
Checks have passed

amicic commented:
We could remove alreadyHaveExclusive flag that was introduced just for this purpose, to help with peculiarities of Metronome.  @gacholio what's your preference?

jdmpapin commented:
I thought that I had deliberately kept `alreadyHaveExclusive`, but I suppose I forgot to. I believe there are other parts of redefinition that could trigger GC, e.g. `recreateRAMClasses()`, `heapifyStackAllocatedObjects()`, though the latter releases and reacquires exclusive VM access, which I imagine it does in order to allow GC to take it. My (unsolicited...) inclination would be to reinstate the line that sets `alreadyHaveExclusive`, and keep it set until redefinition is done (like `isClassUnloadMutexHeldForRedefinition`). And maybe `heapifyStackAllocatedObjects()` could avoid releasing exclusive VM access if `alreadyHaveExclusive` were set beforehand rather than afterward

gacholio commented:
Sure, remove it. It felt like a huge hack, so if we need this in the future we can design a proper solution when not under time pressure.
